File: packages/react-native-v2/android/src/main/java/com/nativeevervault/HybridEvervaultSdk.kt
class HybridEvervaultSdk : HybridEvervaultSdkSpec() {
  private val evervaultScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

  private val instances = mutableMapOf<String, Evervault>()

  private fun getInstanceId(teamId: String, appId: String): String {
    return "$teamId:$appId"
  }

  private fun getInstance(instanceId: String): Evervault? {
    return instances[instanceId]
  }

  override fun initialize(teamId: String, appId: String): String {
    val instance = Evervault(teamId, appId)
    val instanceId = getInstanceId(teamId, appId)
    instances[instanceId] = instance
    return instanceId
  }

  override fun encrypt(instanceId: String, json: String): Promise<String> {
    return Promise.async {
      evervaultScope.launch {
        val instance = getInstance(instanceId)
        if (instance == null) {
          throw Error("Instance not found.")
        }

        val encrypted = withContext(context = Dispatchers.IO) {
          instance.encrypt(json)
        }
        return@async encrypted
      }
    }
  }
}

File: packages/react-native-v2/ios/HybridEvervaultSdk.swift
import Evervault
import NitroModules

class HybridEvervaultSdk : HybridEvervaultSdkSpec {
  var instances: [String: Evervault] = [:]

  func getInstanceId(_ teamId: String, _ appId: String) -> String {
      return "\(teamId):\(appId)"
  }

  func getInstance(_ instanceId: String) -> Evervault? {
      return self.instances[instanceId]
  }

  func initialize(teamId: String, appId: String) -> String {
      let evervault = Evervault(teamId: teamId, appId: appId)
      let instanceId = getInstanceId(teamId, appId)
      self.instances[instanceId] = evervault
      return instanceId
  }

  func encrypt<T, U>(instanceId: String, _ data: T) throws -> Promise<U> {
    return Promise.async {
        guard let instance = self.getInstance(instanceId) else {
              throw RuntimeError.error(withMessage: "Instance not found.")
        }
        return try await instance.encrypt(data) as! U
    }
  }

  func encryptString(instanceId: String, data: String) throws -> Promise<String> {
    return try self.encrypt(instanceId: instanceId, data)
  }

  func encryptNumber(instanceId: String, data: Double) throws -> Promise<String> {
    return try self.encrypt(instanceId: instanceId, data)
  }

  func encryptBoolean(instanceId: String, data: Bool) throws -> Promise<String> {
    return try self.encrypt(instanceId: instanceId, data)
  }

  func encryptObject(instanceId: String, data: AnyMap) throws -> Promise<AnyMap> {
      return try self.encrypt(instanceId: instanceId, data)
  }

  func encryptArray(instanceId: String, data: [AnyMap]) throws -> Promise<[AnyMap]> {
      return try self.encrypt(instanceId: instanceId, data)
  }
}
